### ISSUE TITLE
Handwaving about future usage and outputs

### ISSUE THREAD

**Issue by HealthyPear** [cta-observatory/pyirf#30]
Handwaving based on a discussion with @kosack .
No file formats or data structures, just context and usage possibilities.

**Current situation (+ PR #12 for concrete example):**

_pyirf_ outputs,
- IRFs,
- a table containing information about the cuts applied
- one table per particle type containing the DL2 simulated events which pass said cuts

**Future situation (handwaving):**

We can imagine 2 steps:
1. create performance information (IRFs + applied cuts) from simulated DL2 data
2. apply said cuts to DL2 data (either real or simulated - in the latter case _not_ used to make the IRFs)
3. apply the IRFs to this DL2 selected data to obtain DL3 event lists

Steps 2 & 3 could be performed by a separate tool (my opinion is that _pyirf_ should do it, as it wouldn't need any further information wrt to step 1 from the original pipeline).

About the specific content and usage:

- 1:1:1 correspondence between science case (or in general the user), config file, and DL3 output
- we can imagine 3 types of cuts each with multiple choices, e.g.

| Angular cut  | Background rejection | Quality of reconstruction |
| ------------- | ------------- | ------------- |
| None (aka "full-enclosure")  | loose | best sensitivity |
| point-source (68%, fixed, optimization list)  | standard  | best angular resolution |
| -  | hard | best energy resolution |

so each time we perform step n.1 of the previous section, we are choosing (though the config file) a combination of one value per column, which would result in a unique set of IRFs+cuts to create the correspondent DL3 data.

**Comment by bregeon** [comment]
It seems to me that the tracing of cuts and configuration are a _provenance_ problem.
In principle it would be nice to use the `ctapipe / provenance` module to handle provenance from DL2 to DL3 even in `pyIRF`,
I guess that would solve some of the problems mentioned here.

**Comment by MaxNoe** [comment]
I would probably turn it around.

ctapipe could contain a script using functions from pyirf to produce IRFs and DL3 files that then also have the  correct provenance.

Like this, pyirf would not have to care about provenance, which makes it much simpler!

My vision is that pyirf is a library with the needed functions to optimize cuts for best sensitivity, calculate the resulting IRFs and write them into the DL3 format. 

It should not be concerned with DL2 file formats and the like  but just provide the necessary functions to be called by users.

**Comment by HealthyPear** [comment]
> It seems to me that the tracing of cuts and configuration are a _provenance_ problem.

Yes, sorry I am still not familiar with this nomenclature!

Regarding the ctapipe part:

> ctapipe could contain a script using functions from pyirf to produce IRFs and DL3 files that then also have the correct provenance.

this is exactly why I wanted to separate protopipe.perf from the rest, so the future pipeline module of ctapipe will just call pyirf

 > It should not be concerned with DL2 file formats and the like but just provide the necessary functions to be called by users, which could include ctapipe.

Exactly

**Comment by kosack** [comment]
> It should not be concerned with DL2 file formats and the like but just provide the necessary functions to be called by users, which could include ctapipe.

Yes, this would simplify things a lot. Basically it would be script that takes DL2 inputs and outputs the intermediate products needed by PyIRF .  You can still support reading other non-ctapipe formats, but just make sure to propagate all the headers in the ctapipe format case, so you get all the provenance and other info.

**Comment by HealthyPear** [comment]
Some years passed since this discussion - I think we can close this and open new issues as new ideas pop out. What do you think?

**Comment by kosack** [comment]
yes, this is basically redundant now with event types and there is an open PR for a standard IRF script in ctapipe/datapipe.

**Comment by kosack** [comment]
Though some of the ideas here are still useful, so we could move the issue to the ctapipe tool
